coderd/database/queries/insights.sql

```diff
@@ -329,6 +329,267 @@ SELECT
 FROM unflattened_usage_by_interval
 GROUP BY from_, to_;
 
+-- name: GetTemplateUsageStats :many
+SELECT
+	*
+FROM
+	template_usage_stats
+WHERE
+	start_time >= @start_time::timestamptz
+	AND end_time <= @end_time::timestamptz
+	AND CASE WHEN COALESCE(array_length(@template_ids::uuid[], 1), 0) > 0 THEN template_id = ANY(@template_ids::uuid[]) ELSE TRUE END;
+
+-- name: UpsertTemplateUsageStats :exec
+-- This query aggregates the workspace_agent_stats and workspace_app_stats data
+-- into a single table for efficient storage and querying. Half-hour buckets are
+-- used to store the data, and the minutes are summed for each user and template
+-- combination. The result is stored in the template_usage_stats table.
+WITH
+	latest_start AS (
+		SELECT
+			COALESCE(
+				MAX(start_time) - '1 hour'::interval,
+				NOW() - '6 months'::interval
+			) AS t
+		FROM
+			template_usage_stats
+	),
+	workspace_app_stat_buckets AS (
+		SELECT
+			-- Truncate the minute to the nearest half hour, this is the bucket size
+			-- for the data.
+			date_trunc('hour', s.minute_bucket) + trunc(date_part('minute', s.minute_bucket) / 30) * 30 * '1 minute'::interval AS time_bucket,
+			w.template_id,
+			was.user_id,
+			-- Both app stats and agent stats track web terminal usage, but
+			-- by different means. The app stats value should be more
+			-- accurate so we don't want to discard it just yet.
+			CASE
+				WHEN was.access_method = 'terminal'
+				THEN '[terminal]' -- Unique name, app names can't contain brackets.
```

Comment:
maybe we need a more future proof method to prevent naming conflicts, an extra column for `coder` or `internal` apps?

Reply:
This is going to end up in JSON, e.g. `{"code-server": 30, "[terminal]": 15}`. Adding more text takes up more storage and we already have a protection against conflict, app slugs aren't allowed to contain `[`.